### PR TITLE
Achievements

### PR DESCRIPTION
**Pull request by teodorciuraru**
**Adds achievements to the app.**

Achievements represents titles that the user receives when completing tasks throughout the app.
### **AchievementManager**

Offers methods like `getAchievements(withTag:)`, `unlock(achievement:)` `checkForDuplicates()`, `isAchievementUnlocked(achievement:)`.
### **Specialized achievement trackers (MedicineAchievementsManager, RapidFireAchievementsManager, GeneralAchievementsManager)**

Define and check achievements, overall creating 3 main achievement categories (Pill, Games and General achievements).

_In our developers' opinion, only one more manager could be added (**MythVsFactManager**) at the app's current state. Else, if there are no other games implemented, those 3-4 managers should cover all the achievements that can be invented for this application._
### **Achievement**

Core Data object that defines an achievement for this app.

_Some future achievements could support tracking progress (e.g achieving 2 / 5 of the current achievement's task. If needed, the Achievement class will be modified so that it will also track the progress (maybe another property to do this)_
### **AchievementsViewController**

Represents a View Controller with a Table View that fetches the achievements from all the three categories (**Pills**, **Games**, **General**) and sets the progress bar depending on the achievement's completition.
### **Toasts**

Cocoapod that allows us to present an Android Toast-like alert when the user triggers and achievement.
## Adding achievements
1. Define the Achievement in the corespondent specialized achievement manager (create a new one if adding achievements for the MythVsFact game or another game).
2. Implement a **check_()** method for the achievement in the same manager.
3. Make sure you can use one of the `NSNotificationEvents`, if not create a new observer for the class and the trigger method and post the notification in the code where / when needed.
4. Create unit test for that achievement.
## Additional fix

This PR also adds a `LocationTextLabel` validation so that a trip could not be generated without a location name.

### CODE REVIEW

File: malaria-ios/Achievement+CoreDataProperties.swift [hunk 1]
```diff
@@ -0,0 +1,22 @@
+//
+//  Achievement+CoreDataProperties.swift
+//  malaria-ios
+//
+//  Created by Teodor Ciuraru on 7/28/16.
```
Comment (bphenriques) [inline malaria-ios/Achievement+CoreDataProperties.swift]: Remove

File: malaria-ios/Achievement+CoreDataProperties.swift [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+//  Copyright © 2016 Bruno Henriques. All rights reserved.
+//
+//  Choose "Create NSManagedObject Subclass…" from the Core Data editor menu
+//  to delete and recreate this implementation file for your updated model.
+//
+
+import Foundation
+import CoreData
+
+extension Achievement {
```
Comment (bphenriques) [inline malaria-ios/Achievement+CoreDataProperties.swift]: This class is generated by the model, it shouldn't be an extension.

File: malaria-ios/Achievement.swift [hunk 3]
```diff
@@ -0,0 +1,37 @@
+import Foundation
+import CoreData
+
+/// A general feat that a user can unlock using the application thoroughly.
+
+class Achievement: NSManagedObject {
```
Comment (bphenriques) [inline malaria-ios/Achievement.swift]: This is the one who should be an extension

File: malaria-ios/Managers/AchievementManager.swift [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      return
+    }
+    
+    // Unlock achievement.
+    achievement.isUnlocked = true
+    
+    Logger.Info("Unlocked achievement: \(name)")
+    
+    // Show alert message.
+    UIApplication.sharedApplication().keyWindow?.rootViewController?
```
Comment (bphenriques) [inline malaria-ios/Managers/AchievementManager.swift]: Make a global function for the toasts (reduces boiler plate code and it's easy to reuse in other places of the code).

``` swift
func createToast(text: string, duration: int. ...) {
UIApplication.sharedApplication().keyWindow?.rootViewController?.view.makeToast(...)
}
```

If it is possible to provide a image optionally better (for medicine achievements).

File: malaria-ios/Managers/GeneralAchievementsManager.swift [hunk 5]
```diff
@@ -0,0 +1,60 @@
+import Foundation
+
+/// Class that handles all the achievements related to the Rapid Fire Game.
```
Comment (bphenriques) [inline malaria-ios/Managers/GeneralAchievementsManager.swift]: Comment outdated?

File: malaria-ios/Managers/GeneralAchievementsManager.swift [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+class GeneralAchievementsManager: NSObject {
```
Comment (bphenriques) [inline malaria-ios/Managers/GeneralAchievementsManager.swift]: I don't follow this. In your description you describe "RapidFireAchievementsManager", "MedicineAchievementsManager", but it seems that this handles Trips. If so, rename the class...

File: malaria-ios/Managers/GeneralAchievementsManager.swift [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+
+// MARK: Checking Achievements
+
+extension GeneralAchievementsManager {
+  
+  func checkAchievements() {
+    checkFlawlessGame()
+  }
+  
+  private func checkFlawlessGame() {
```
Comment (bphenriques) [inline malaria-ios/Managers/GeneralAchievementsManager.swift]: This is for trips... not for games. Rename. CheckFirstTripDefined or something

File: malaria-ios/Managers/RapidFireAchievementManager.swift [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+class RapidFireAchievementManager: NSObject {
+  static let sharedInstance = RapidFireAchievementManager()
+  
+  private var game: RapidFireGame?
+  private let achievementManager = AchievementManager.sharedInstance
+  
+  override init() {
+    super.init()
+    
+    defineAchievements()
```
Comment (bphenriques) [inline malaria-ios/Managers/RapidFireAchievementManager.swift]: Refactor:
To force developers to do this right. You can define a protocol or a abstract class where your achivementsManager have to define "defineAchivements" you might can refactor and provide to the NSNotificationEvents argument to the super class and force the classes to have a "checkAchivement" argument. Is this clear to you? TL ; DR; Abstract class that receives the NSNotificationEvent with defineAchivement and checkAchivement as abstract methods.

File: malaria-ios/ViewControllers/AchievementsViewController.swift [hunk 9]
```diff
@@ -1,10 +1,98 @@
...
+    super.viewWillAppear(animated)
+    updateModel()
+  }
+  
+  func updateModel() {
+    allAchievements = []
+    
+    let achievementManager = AchievementManager.sharedInstance
+    
+    let gamesAchievements =
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/AchievementsViewController.swift]: In my opinion it is better to
let pillAchivements = ...
appendToModel( ... )

it is easier to detect the 2 lines pattern for more achivements. Please change accordingly

File: malaria-ios/ViewControllers/AchievementsViewController.swift [hunk 10]
```diff
@@ -1,10 +1,98 @@
...
+    return allAchievements.count
+  }
+  
+  func tableView(tableView: UITableView, titleForHeaderInSection section: Int) -> String? {
+    return allAchievements[section].sectionName
+  }
+  
+  func tableView(tableView: UITableView, cellForRowAtIndexPath indexPath: NSIndexPath) -> UITableViewCell {
+    
+    let cell = tableView.dequeueReusableCellWithIdentifier("Achievement Cell", forIndexPath: indexPath) as! AchievementTableViewCell
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/AchievementsViewController.swift]: move cell identifier to a class constant

File: malaria-ios/ViewControllers/AchievementsViewController.swift [hunk 11]
```diff
@@ -1,10 +1,98 @@
...
+
+extension AchievementsViewController: UITableViewDelegate {
+  
+  func tableView(tableView: UITableView, willDisplayHeaderView view: UIView, forSection section: Int) {
+    
+    // This changes the header background
+    view.tintColor = Constants.DefaultGreenTint
+    
+    // Gets the header view as a UITableViewHeaderFooterView and changes the text color
+    let headerView: UITableViewHeaderFooterView = view as! UITableViewHeaderFooterView
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/AchievementsViewController.swift]: ? Separate into different lines

File: malaria-ios/ViewControllers/Constants.swift [hunk 12]
```diff
@@ -2,25 +2,44 @@ import UIKit
...
     static let PillStats = "PillScreens"
     static let InfoHub = "InfoHub"
   }
   
   struct RapidFireGame {
-    
-    // How rapidly should the Remaining Seconds blink
-    // measured in seconds.
+    // How rapidly should the Remaining Seconds blink measured in seconds.
     static let RemainingSecondsBlinkRate = 0.2
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/Constants.swift]: If it is not shared in the app, move to the RapidFireGame class constant. I think this is used in the view controller, right?

File: malaria-ios/ViewControllers/Constants.swift [hunk 13]
```diff
@@ -2,25 +2,44 @@ import UIKit
...
+  
+  struct Achievements {
+    struct Tags {
+      static let Pills = "Pills"
+      static let Games = "Games"
+      static let General = "General"
+    }
+    
+    struct Pills {
+      static let StayingSafe = "Staying safe"
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/Constants.swift]: This should be define in the Manager class not here.

File: malaria-ios/ViewControllers/Constants.swift [hunk 14]
```diff
@@ -2,25 +2,44 @@ import UIKit
...
+    }
+    
+    struct RapidFireGame {
+      static let FlawlessGame = "Flawless game"
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/Constants.swift]: ditto.

File: malaria-ios/ViewControllers/Constants.swift [hunk 15]
```diff
@@ -2,25 +2,44 @@ import UIKit
...
+    }
+    
+    struct General {
+      static let PlanFirstTrip = "Plan First Trip"
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/Constants.swift:22]: ditto.

File: malaria-ios/ViewControllers/Constants.swift [hunk 16]
```diff
@@ -2,25 +2,44 @@ import UIKit
...
-    
-    // How rapidly should the Remaining Seconds blink
-    // measured in seconds.
+    // How rapidly should the Remaining Seconds blink measured in seconds.
     static let RemainingSecondsBlinkRate = 0.2
   }
-}
+  
+  struct Achievements {
+    struct Tags {
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/Constants.swift]: Each manager should be able to define their own Tag? The view controller can access the existing tags and not controlling them. Moreover, it removes this struct from here.

File: malaria-ios/ViewControllers/PlanTrip/PlanTripViewController.swift [hunk 17]
```diff
@@ -151,6 +151,16 @@ extension PlanTripViewController{
     }
     
     @IBAction func generateTrip(sender: AnyObject) {
+        if location.text?.characters.count == 0 {
+          
+          // Show alert message.
+          UIApplication.sharedApplication().keyWindow?.rootViewController?
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/PlanTrip/PlanTripViewController.swift]: ditto

File: malaria-iosTests/TestAchievementsManager.swift [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+  
+  func testGetAchievements() {
+    let tag = "Test tag"
+    
+    Achievement.define("Test achievement", description: "Test",
+                                         tag: tag)
+    
+    let achievements = am.getAchievements(withTag: tag)
+
+    XCTAssert(achievements.count == 1)
```
Comment (bphenriques) [inline malaria-iosTests/TestAchievementsManager.swift]: Add tests getAchivements, with undefined tags
Add tests getAchievements with multiple achivements per one tag
Add tests getAchivements with 2 tags with multiple achivements and test them individually.

File: malaria-ios/ViewControllers/PlanTrip/PlanTripViewController.swift [hunk 19]
```diff
@@ -175,14 +185,16 @@ extension PlanTripViewController{
...
+    private func storeTrip() {
         let trip = tripsManager.createTrip(location.text!, medicine: medicine.name(), departure: departureDay, arrival: arrivalDay, timeReminder: reminder)
         let itemManager = trip.itemsManager
         items.foreach({ itemManager.addItem($0.0, quantity: 1) })
         itemManager.toggleCheckItem( items.filter({ $0.1 }).map({ $0.0 }))
         
         scheduleNotifications(trip)
         prepareHistoryValuePicker()
+      
+        NSNotificationEvents.TripPlanned()
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/PlanTrip/PlanTripViewController.swift]: This should be inside scheduleNotifications (backend)

File: malaria-iosTests/TestGeneralAchievements.swift [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+  }
+  
+  func testFirstTrip() {
+    var achievementUnlocked = am.isAchievementUnlocked(achievement: Constants.Achievements.General.PlanFirstTrip)
+    
+    XCTAssertFalse(achievementUnlocked)
+
+    tm.createTrip("dasda", medicine: "", departure: NSDate(), arrival: NSDate().dateByAddingTimeInterval(30000), timeReminder: NSDate().dateByAddingTimeInterval(2000))
+    
+    NSNotificationEvents.TripPlanned()
```
Comment (bphenriques) [inline malaria-iosTests/TestGeneralAchievements.swift]: TripPlanned should be defined in the backend.

File: malaria-iosTests/TestGeneralAchievements.swift [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+    super.tearDown()
+    am.clearAchievements()
+  }
+  
+  func testFirstTrip() {
+    var achievementUnlocked = am.isAchievementUnlocked(achievement: Constants.Achievements.General.PlanFirstTrip)
+    
+    XCTAssertFalse(achievementUnlocked)
+
+    tm.createTrip("dasda", medicine: "", departure: NSDate(), arrival: NSDate().dateByAddingTimeInterval(30000), timeReminder: NSDate().dateByAddingTimeInterval(2000))
```
Comment (bphenriques) [inline malaria-iosTests/TestGeneralAchievements.swift]: let tripDate = NSDate()
let arrival = tripDate + 30.day
let reminder = arrival - 1.day

Use the existing NSDate extensions

File: malaria-iosTests/TestGeneralAchievements.swift [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+    
+    NSNotificationEvents.TripPlanned()
+
+    print(tm.getTrip() != nil ? "cioc" : "boc")
```
Comment (bphenriques) [inline malaria-iosTests/TestGeneralAchievements.swift]: remove

File: malaria-iosTests/TestMedicineAchievements.swift [hunk 23]
```diff
@@ -0,0 +1,45 @@
...
+    m.clearCoreData()
+  }
+  
+  func testStayingSafeAchievement() {
+    var achievementUnlocked = am.isAchievementUnlocked(achievement: Constants.Achievements.Pills.StayingSafe)
+    
+    XCTAssertFalse(achievementUnlocked)
+    
+    // Take medicine for today
+
```
Comment (bphenriques) [inline malaria-iosTests/TestMedicineAchievements.swift]: remove line (l:38)

File: malaria-iosTests/TestMedicineAchievements.swift [hunk 24]
```diff
@@ -0,0 +1,45 @@
...
+  
+  override func tearDown() {
+    super.tearDown()
+    am.clearAchievements()
+    m.clearCoreData()
+  }
+  
+  func testStayingSafeAchievement() {
+    var achievementUnlocked = am.isAchievementUnlocked(achievement: Constants.Achievements.Pills.StayingSafe)
+
```
Comment (bphenriques) [inline malaria-iosTests/TestMedicineAchievements.swift]: remove line (34)

File: malaria-iosTests/TestMedicineAchievements.swift [hunk 25]
```diff
@@ -0,0 +1,45 @@
...
+    var achievementUnlocked = am.isAchievementUnlocked(achievement: Constants.Achievements.Pills.StayingSafe)
+    
+    XCTAssertFalse(achievementUnlocked)
+    
+    // Take medicine for today
+    
+    m.getCurrentMedicine()!.registriesManager.addRegistry(NSDate(), tookMedicine: true, modifyEntry: true)
+    
+    achievementUnlocked = am.isAchievementUnlocked(achievement: Constants.Achievements.Pills.StayingSafe)
+
```
Comment (bphenriques) [inline malaria-iosTests/TestMedicineAchievements.swift]: remove line 42

File: malaria-iosTests/TestMedicineManager.swift [hunk 26]
```diff
@@ -26,7 +26,7 @@ class TestSetupInsertClear: XCTestCase {
     func testCurrentMedicine(){
         if let medi = m.getCurrentMedicine(){
             XCTAssertEqual(medi.name, Medicine.Pill.Malarone.name())
-            XCTAssertEqual(medi.interval, 1)
+            XCTAssertEqual(medi.internalInterval, 1)
```
Comment (bphenriques) [inline malaria-iosTests/TestMedicineManager.swift]: Why this change?

InternalInterval is private
interval is public which converts internalInterval to Int

Revert accordingly

File: malaria-iosTests/TestMedicineManager.swift [hunk 27]
```diff
@@ -36,7 +36,7 @@ class TestSetupInsertClear: XCTestCase {
     func testGetMedicine(){
         if let medi = m.getMedicine(currentPill.name()){
             XCTAssertEqual(medi.name, Medicine.Pill.Malarone.name())
-            XCTAssertEqual(medi.interval, 1)
+            XCTAssertEqual(medi.internalInterval, 1)
```
Comment (bphenriques) [inline malaria-iosTests/TestMedicineManager.swift]: ditto

File: malaria-ios/Achievement+CoreDataProperties.swift (hunk 2 above)
Comment (teodorciuraru) [inline malaria-ios/Achievement+CoreDataProperties.swift]: No issue here, right?

File: malaria-ios/Achievement.swift (hunk 3 above)
Comment (teodorciuraru) [inline malaria-ios/Achievement.swift]: Ditto

File: malaria-ios/Achievement+CoreDataProperties.swift (hunk 2 above)
Comment (bphenriques) [inline malaria-ios/Achievement+CoreDataProperties.swift]: Nop. Xcode tricked me. Remove the "Created by"...

File: malaria-ios/Achievement.swift (hunk 3 above)
Comment (bphenriques) [inline malaria-ios/Achievement.swift]: Ditto.

File: malaria-ios/ViewControllers/AchievementsViewController.swift (hunk 9 above)
Comment (teodorciuraru) [inline malaria-ios/ViewControllers/AchievementsViewController.swift]: I don't understand well. Can you please elaborate?

File: malaria-ios/ViewControllers/AchievementsViewController.swift (hunk 11 above)
Comment (teodorciuraru) [inline malaria-ios/ViewControllers/AchievementsViewController.swift]: I deleted the type, it should be inferred.

File: malaria-ios/ViewControllers/AchievementsViewController.swift (hunk 9 above)
Comment (bphenriques) [inline malaria-ios/ViewControllers/AchievementsViewController.swift]: Group together:

``` swift
 let manager1 = (...)
appendToModel(...)

let manager2 = (...)
appendToModel(...)
```

Instead of:

``` swift
let manager1 = (...)
let manager2 = (...)

appendToModel(...)
appendToModel(...)
```

File: malaria-iosTests/TestGeneralAchievements.swift (hunk 22 above)
Comment (teodorciuraru) [inline malaria-iosTests/TestGeneralAchievements.swift]: You didn't see a thing.

File: malaria-ios/ViewControllers/Constants.swift (hunk 15 above)
Comment (teodorciuraru) [inline malaria-ios/ViewControllers/Constants.swift:22]: I tried removing the constants, but I use them in the tests as well. I need them in this file.

File: malaria-iosTests/TestMedicineAchievements.swift (hunk 24 above)
Comment (teo029) [inline malaria-iosTests/TestMedicineAchievements.swift]: If I remove it, I should remove in other places as well.

Comment (teo029) [inline malaria-iosTests/TestMedicineAchievements.swift]: Does it look that bad?

File: malaria-iosTests/TestMedicineManager.swift (hunk 26 above)
Comment (teodorciuraru) [inline malaria-iosTests/TestMedicineManager.swift]: By mistake. Nicely stated!

File: malaria-ios/ViewControllers/Constants.swift (hunk 16 above)
Comment (teodorciuraru) [inline malaria-ios/ViewControllers/Constants.swift]: Can you please elaborate, I didn't understand.

File: malaria-iosTests/TestAchievementsManager.swift (hunk 18 above)
Comment (teodorciuraru) [inline malaria-iosTests/TestAchievementsManager.swift]: I don't get the last one. Achievements might only have one tag. What am I missing?

Comment (bphenriques) [inline malaria-iosTests/TestAchievementsManager.swift]: Try to add two achivements for one tag and check if they were correctly stored.

File: malaria-ios/ViewControllers/Constants.swift (hunk 16 above)
Comment (bphenriques) [inline malaria-ios/ViewControllers/Constants.swift]: I am looking at the constants and I think you can move the constants (might) to each Manager. Afterwards, the view controller merges the existing tags defined in each manager. At the moment, the view controller depends on those constants defined in the "Constants" file. It might be hard to explain, we can meet later today.

File: malaria-ios/Managers/Achievements/AchievementManager.swift [hunk 28]
```diff
@@ -0,0 +1,82 @@
...
+/// Handles general achievement logic throughout the app.
+
+class AchievementManager: CoreDataContextManager {
+  static var sharedInstance = AchievementManager(
+    context: CoreDataHelper.sharedInstance.createBackgroundContext()!
+  )
+  
+  typealias AchievementObject = (sectionName: String, achievements: [Achievement])
+  
+  private var tags: [String]? = []
```
Comment (bphenriques) [inline malaria-ios/Managers/Achievements/AchievementManager.swift]: why optional?

File: malaria-ios/Managers/Achievements/AchievementManager.swift [hunk 29]
```diff
@@ -0,0 +1,82 @@
...
+  
+  // Check if the achievement was already unlocked and return.
+  
+  func isAchievementUnlocked(achievement name: String) -> Bool {
+    guard let achievement = achievements.filter({ $0.name == name }).first else {
+      Logger.Error("Tried to unlock an achievement that doesn't exist.")
+      return false
+    }
+    
+    return achievement.isUnlocked == true
```
Comment (bphenriques) [inline malaria-ios/Managers/Achievements/AchievementManager.swift]: return directly the value... :-P

File: malaria-ios/Managers/Achievements/Specialized Achievement Managers/Games/RapidFireAchievementManager.swift [hunk 30]
```diff
@@ -0,0 +1,69 @@
...
+  private let achievementManager = AchievementManager.sharedInstance
+  
+  let tag = "Games"
+    
+  override init() {
+    super.init()
+    
+    defineAchievements()
+    
+    achievementManager.addTag(tag)
```
Comment (bphenriques) [inline malaria-ios/Managers/Achievements/Specialized Achievement Managers/Games/RapidFireAchievementManager.swift]: Move the addTag(tag) to achivement.Define

File: malaria-ios/Managers/Achievements/Specialized Achievement Managers/Medicine /MedicineAchievementManager.swift [hunk 31]
```diff
@@ -0,0 +1,72 @@
...
+
+  let tag = "Pills"
+
+  init(medicine: Medicine) {
+    self.medicine = medicine
+    self.context = medicine.managedObjectContext
+    
+    super.init()
+    
+    achievementManager.addTag(tag)
```
Comment (bphenriques) [inline malaria-ios/Managers/Achievements/Specialized Achievement Managers/Medicine /MedicineAchievementManager.swift]: ditto. Reason: the tags have to be defined anyway therefore it might as well spare this line of code throughout the app by defining directly in the achivement.define method

File: malaria-ios/Managers/TripHelpers/TripsManager.swift [hunk 32]
```diff
@@ -1,90 +1,92 @@
...
+  func getHistory(limit: Int = 15) -> [TripHistory] {
+    return TripHistory.retrieve(TripHistory.self, fetchLimit: limit, context: context).sort({ $0.timestamp > $1.timestamp})
+  }
+  
+  /// Appends the trip to the history
+  ///
+  /// - parameter `Trip`:: the trip
+  private func createHistory(trip: Trip){
+    let previousHistory = getHistory(Int.max)
+    if previousHistory.count >= 15 {
```
Comment (bphenriques) [inline malaria-ios/Managers/TripHelpers/TripsManager.swift]: Argh, my fault... move to class constant

File: malaria-ios/ViewControllers/Info Hub and Games/Games Lists/Achievements/AchievementTableViewCell.swift [hunk 33]
```diff
@@ -0,0 +1,24 @@
...
+  
+  private let MaximumProgress: Float = 1.0
+  
+  func updateCell(achievement: Achievement) {
+    
+    achievementName.text = achievement.name
+    achievementDescription.text = achievement.desc
+    
+    // Check if achievement is unlocked.
+    if achievement.isUnlocked == true {
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/Info Hub and Games/Games Lists/Achievements/AchievementTableViewCell.swift]: == true is redundant.

File: malaria-ios/ViewControllers/Info Hub and Games/Games Lists/Achievements/AchievementTableViewCell.swift [hunk 34]
```diff
@@ -0,0 +1,24 @@
...
+      progressView.setProgress(MaximumProgress, animated: true)
+      progressLabel.text = "Completed"
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/Info Hub and Games/Games Lists/Achievements/AchievementTableViewCell.swift]: Move label to a constant

File: malaria-ios/ViewControllers/Info Hub and Games/Games Lists/Achievements/AchievementTableViewCell.swift [hunk 35]
```diff
@@ -0,0 +1,24 @@
...
+      return
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/Info Hub and Games/Games Lists/Achievements/AchievementTableViewCell.swift]: return is redundant

File: malaria-ios/ViewControllers/Info Hub and Games/Games Lists/Games/Myth vs. Fact/MythVsFactViewController.swift [hunk 36]
```diff
@@ -70,6 +68,14 @@ class MythVsFactViewController: GameViewController {
     statementLabelInitialFrame = statementLabel.frame
   }
   
+  override func endGame() {
+    super.endGame()
+    
+    // Send a notification that the game has finished.
+    let dict = ["game" : mythVsFactGame]
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/Info Hub and Games/Games Lists/Games/Myth vs. Fact/MythVsFactViewController.swift:75]: The "game" is meant to be global for each game?

This seems like a candidate to a class `let gameFinishedEventArgs = ["game" : ... ]`

File: malaria-ios/ViewControllers/Info Hub and Games/Games Lists/Games/Rapid Fire/RapidFireViewController.swift [hunk 37]
```diff
@@ -73,8 +72,16 @@ class RapidFireViewController: GameViewController {
   }
   
   override func endGame() {
-    super.endGame()
+    // Save the user's final score
+    game.maximumScore = userScore
+    
+    // Send a notification that the game has finished.
+    let dict = ["game": rapidFireGame]
```
Comment (bphenriques) [inline malaria-ios/ViewControllers/Info Hub and Games/Games Lists/Games/Rapid Fire/RapidFireViewController.swift:79]: ok. I undestand what you are doing here. Do you need to pass the dictionary? Can you simplify?

Want to understand :)

Comment (teodorciuraru) [inline malaria-ios/ViewControllers/Info Hub and Games/Games Lists/Games/Rapid Fire/RapidFireViewController.swift:79]: This is the correct way to do it as for Swift 2. Couldn't find a better method.

File: malaria-ios/ViewControllers/Info Hub and Games/Games Lists/Games/Myth vs. Fact/MythVsFactViewController.swift (hunk 36 above)
Comment (teodorciuraru) [inline malaria-ios/ViewControllers/Info Hub and Games/Games Lists/Games/Myth vs. Fact/MythVsFactViewController.swift:75]: Game could be global for all games, but there I just needed a name.

I don't understand the candidate part. Do you think a better design can be added?

File: malaria-ios/ViewControllers/Info Hub and Games/Games Lists/Games/Rapid Fire/RapidFireViewController.swift (hunk 37 above)
Comment (bphenriques) [inline malaria-ios/ViewControllers/Info Hub and Games/Games Lists/Games/Rapid Fire/RapidFireViewController.swift:79]: Discussed, you are right ;)
